### PR TITLE
Presentation mode

### PR DESCRIPTION
**Pull request by ronak0624**
Transitions in presentation upload to firebase with slide number, previous slide, titles, and timestamps. To view the info locally, uncomment line 168 in yoodli-chrome-extension-react/src/pages/background/index.js, and view the Background page console from the 'Manage Extensions' page on Chrome. 

Before this gets merged, the yoodli-client TranscriptionSection needs to have some logic changed. When I record in only PresentationMode or only in regular editing mode, everything works as it should, but if I transition between both modes in the same recording, the transcription section will crash that section with this error. Some of the chunk parsing on that page is wonky with switching modes. 

Uncaught TypeError: Cannot read properties of undefined (reading 'startTimestamp')
    at TranscriptSection.js:154
    at TranscriptSection.js:39

### CODE REVIEW

**Comment by eshajoshi** [comment]
Hey @ronak0624, nice work on the presentation mode slide detection feature! I see that you're writing an array of objects called `presentationModelTransitions` to Firestore. The good news is that it's looking accurate (I've tested only a few times!). Currently TranscriptSection on the client is expecting a field named `slideChangeUrls`. Could we have the chrome extension write only 1 field to Firestore? It doesn't matter whether the Slides were presented in non presenter mode, presenter mode, or both. But regardless, those `from` and `to` values and `timeMs` values would captured. How does this sound?

File: yoodli-chrome-extension-react/src/pages/background/index.js [hunk 1]
```diff
@@ -152,6 +153,7 @@ function cancelScreenSharing(_msg) {
             const additionalAnalysisData = {}
             if (tabChangeUrls.length) additionalAnalysisData.tabChangeUrls = tabChangeUrls 
             if (slideChangeUrls.length) additionalAnalysisData.slideChangeUrls = slideChangeUrls
+            if (presentationModeTransitions.length) additionalAnalysisData.presentationModeTransitions = presentationModeTransitions
```
Comment (eshajoshi) [inline yoodli-chrome-extension-react/src/pages/background/index.js:156]: Ideally, the chrome extension would send 1 object representing the slide changes - either slideChangeUrls or presentationModeTransitions but not both. The object would capture the slide changes regardless of the mode the slides were presented in: in pres mode, not in pres mode, or both (both is the common scenario)

Comment (ronak0624) [reply yoodli-chrome-extension-react/src/pages/background/index.js:156]: I'll merge the two arrays together by increasing timestamp

File: yoodli-chrome-extension-react/src/pages/background/index.js [hunk 2]
```diff
@@ -273,6 +275,16 @@ chrome.runtime.onConnect.addListener(port => {
                 activeUrl = tabs[0].url
             })
         }
+        if(msg.type === 'SLIDE_CHANGE') {
```
[the remark this answers is not in the document]
Comment (ronak0624) [reply yoodli-chrome-extension-react/src/pages/background/index.js:278]: Should I put all the other msg.types that David implemented into enums as well?

File: yoodli-chrome-extension-react/src/listeners/presentationListener.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            if (doc.readyState === "complete") {
+                ready.call(doc);
+            } else {
+
+                addEvent(doc, "DOMContentLoaded", ready);
+                addEvent(doc, "readystatechange", readyState);
+            }
+        } else {
+
+            timer = setTimeout(checkLoaded, 1);
```
Comment (dhlanm) [inline yoodli-chrome-extension-react/src/listeners/presentationListener.js:47]: not a huge deal here, but in general I'm leery of having timeout = 1ms unless there's a pretty strong reason to have it

File: yoodli-chrome-extension-react/src/manifest.json [hunk 4]
```diff
@@ -16,6 +16,17 @@
       "js": [
         "overlay.bundle.js"
       ]
+    },
+    {
+      "matches": [
+        "http://docs.google.com/*",
+        "https://docs.google.com/*"
```
Comment (dhlanm) [inline yoodli-chrome-extension-react/src/manifest.json:23]: nice job limiting scope here

**Review by dhlanm — APPROVED** [review]

File: yoodli-chrome-extension-react/src/pages/background/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (eshajoshi) [reply yoodli-chrome-extension-react/src/pages/background/index.js:278]: Just seeing this! Ideally yes, but let's have you please do this in your next chrome ext PR so I'm not blocking you here!

**Comment by eshajoshi** [comment]
Hey @ronak0624 , nice work on this!

I tested this with the yoodli-client `ronak-test-pres-mode` branch and got the following Speakeasy speech (on the PROD db):

http://localhost:3000/app/journal/i3I4zTUQFiaV7g3SZdbAP6VxNMB2/feedback/iivrOZJOsbdegu6nefgz

https://www.projectspeakeasy.com/app/journal/i3I4zTUQFiaV7g3SZdbAP6VxNMB2/feedback/iivrOZJOsbdegu6nefgz

Currently it's not capturing the slide changes when I'm in **edit** mode, though it's doing a good job capturing them when I'm in presenter mode! Could you please look into how to capture it in edit mode too?

On the chrome extension, check out the actual assigned var of _slideChangeUrls_ (that David implemented, not the one being sent into Firestore). That gives us the information we want... we somehow have to merge them with _presentationModeTransitions_

**Comment by ronak0624** [comment]
Thanks @eshajoshi 

> http://localhost:3000/app/journal/i3I4zTUQFiaV7g3SZdbAP6VxNMB2/feedback/iivrOZJOsbdegu6nefgz
> 
> https://www.projectspeakeasy.com/app/journal/i3I4zTUQFiaV7g3SZdbAP6VxNMB2/feedback/iivrOZJOsbdegu6nefgz
> 
> Currently it's not capturing the slide changes when I'm in **edit** mode, though it's doing a good job capturing them when I'm in presenter mode! Could you please look into how to capture it in edit mode too?
> 

I haven't been able to reproduce this, the videos seem like the functionality from before the last 5 commits I have made. A refresh of `npm run watch` -> refresh extension in chrome -> refresh slides will hopefully do the trick, that reload flow is super buggy all the time and I'm willing to bet its just stuck on an old cached version. 

Here's a recording from just now, https://www.projectspeakeasy.com/app/journal/Tj9O4B3iMYg6ktZpbqhh67HaPHr1/feedback/rO3CQLk75z16X40vMRIn
slide numbers are wrong since the titles i added aren't merged yet

> On the chrome extension, check out the actual assigned var of _slideChangeUrls_ (that David implemented, not the one being sent into Firestore). That gives us the information we want... we somehow have to merge them with _presentationModeTransitions_

It is actually only using 1 array to track both edit mode and slide presentation mode. There are 2 listeners that alternate depending on the mode.

**Comment by eshajoshi** [comment]
Nice work!

**Pull request merged**